File: app/src/main/java/net/mmho/photomap2/DistanceActionProvider.java
package net.mmho.photomap2;

import android.content.Context;
import android.support.v4.view.ActionProvider;
import android.view.MenuItem;
import android.view.SubMenu;
import android.view.View;

public class DistanceActionProvider extends ActionProvider
    implements MenuItem.OnMenuItemClickListener{
    private OnDistanceChangeListener onDistanceChangeListener;
    private int selected;

    private static final int INITIAL_INDEX = 1;

    private static final int[] distance={
            7,6,5,4,3

    };
    private static final int[] meters={
            150,
            600,
            5000,
            20000,
            150000,
    };

    private String pretty(int position) {
        int meter = meters[position];
        String unit = "m";
        if(meter>=1000){
            unit = "km";
            meter /= 1000;
        }
        return getContext().getString(R.string.about)+String.valueOf(meter)+unit;
    }


    public DistanceActionProvider(Context context) {
        super(context);
        selected = INITIAL_INDEX;
    }

    public static int initialIndex(){
        return INITIAL_INDEX;
    }

    public void setDistanceIndex(int index){
        selected = index;
    }

    public static int getDistance(int index){
        return distance[index];
    }

    @Override
    public View onCreateActionView() {
        return null;
    }

    @Override
    public boolean hasSubMenu() {
        return true;
    }

    @Override
    public void onPrepareSubMenu(SubMenu subMenu) {
        subMenu.clear();

        for(int i=0;i<distance.length;i++){
            MenuItem s = subMenu.add(0, i, i, pretty(i));
            s.setOnMenuItemClickListener(this);
            if(i==selected) s.setEnabled(false);
        }
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        selected = item.getItemId();
        if(onDistanceChangeListener !=null) onDistanceChangeListener.onDistanceChange(selected);
        return false;
    }
    
    public interface OnDistanceChangeListener {
        void onDistanceChange(int index);
    }

    public void setOnDistanceChangeListener(OnDistanceChangeListener changeListener){
        onDistanceChangeListener = changeListener;
    }

}

File: app/src/main/kotlin/net/mmho/photomap2/DistanceActionProvider.kt
package net.mmho.photomap2

import android.content.Context
import android.support.v4.view.ActionProvider
import android.view.MenuItem
import android.view.SubMenu
import android.view.View

class DistanceActionProvider(context: Context) : ActionProvider(context), MenuItem.OnMenuItemClickListener {
    private var onDistanceChangeListener: OnDistanceChangeListener? = null
    private var selected: Int = 0

    private fun pretty(position: Int): String {
        var meter = meters[position]
        var unit = "m"
        if (meter >= 1000) {
            unit = "km"
            meter /= 1000
        }
        return context.getString(R.string.about) + meter.toString() + unit
    }


    init {
        selected = INITIAL_INDEX
    }

    fun setDistanceIndex(index: Int) {
        selected = index
    }

    override fun onCreateActionView(): View? {
        return null
    }

    override fun hasSubMenu(): Boolean {
        return true
    }

    override fun onPrepareSubMenu(subMenu: SubMenu?) {
        subMenu!!.clear()

        for (i in distance.indices) {
            val s = subMenu.add(0, i, i, pretty(i))
            s.setOnMenuItemClickListener(this)
            if (i == selected) s.isEnabled = false
        }
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        selected = item.itemId
        onDistanceChangeListener?.onDistanceChange(selected)
        return false
    }

    interface OnDistanceChangeListener {
        fun onDistanceChange(index: Int)
    }

    fun setOnDistanceChangeListener(changeListener: OnDistanceChangeListener) {
        onDistanceChangeListener = changeListener
    }

    companion object {

        private val INITIAL_INDEX = 1

        private val distance = intArrayOf(7, 6, 5, 4, 3)
        private val meters = intArrayOf(150, 600, 5000, 20000, 150000)

        fun initialIndex(): Int {
            return INITIAL_INDEX
        }

        fun getDistance(index: Int): Int {
            return distance[index]
        }
    }

}
